File: services/payment-svc/src/api/v1/health.ts
/**
 * Health probes for Kubernetes liveness / readiness.
 */

import type { FastifyInstance } from 'fastify';

import { getConfig } from '@/config/index.js';
import { getRedis } from '@/clients/redis.js';
import { getPool } from '@/db/pool.js';

export async function registerHealthRoutes(app: FastifyInstance): Promise<void> {
  app.get('/livez', async () => {
    const cfg = getConfig();
    return { status: 'ok', version: cfg.APP_VERSION };
  });

  app.get('/readyz', async (_request, reply) => {
    const checks: Record<string, string> = {};
    let overall = true;
    try {
      await getPool().query('SELECT 1');
      checks.postgres = 'ok';
    } catch (err) {
      checks.postgres = `down: ${(err as Error).message}`;
      overall = false;
    }
    try {
      await getRedis().ping();
      checks.redis = 'ok';
    } catch (err) {
      checks.redis = `down: ${(err as Error).message}`;
      overall = false;
    }
    if (!overall) {
      reply.code(503);
    }
    return { ok: overall, checks };
  });
}

File: services/bounty-svc/app/api/v1/health.py
"""Health probes."""

from __future__ import annotations

from fastapi import APIRouter, Depends, Request, Response, status
from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.session import get_session

router = APIRouter(tags=["health"])


@router.get("/livez")
async def livez(request: Request) -> dict:
    version = getattr(request.app.state, "version", "unknown")
    return {"status": "ok", "version": version}


@router.get("/readyz")
async def readyz(
    request: Request,
    response: Response,
    session: AsyncSession = Depends(get_session),
) -> dict:
    checks: dict[str, str] = {}
    overall = True
    try:
        await session.execute(text("SELECT 1"))
        checks["postgres"] = "ok"
    except Exception as e:
        checks["postgres"] = f"down: {e}"
        overall = False
    redis = getattr(request.app.state, "redis", None)
    if redis:
        try:
            await redis.ping()
            checks["redis"] = "ok"
        except Exception as e:
            checks["redis"] = f"down: {e}"
            overall = False
    if not overall:
        response.status_code = status.HTTP_503_SERVICE_UNAVAILABLE
    return {"ok": overall, "checks": checks}
